### PR TITLE
Fix auto-registeration of authN and authZ middlewares

### LINKED ISSUE #42240: Implicitly added AuthN/AuthZ middleware leads to unexpected behavior
In exploring some AuthZ ideas/patterns using the new `WebApplicationBuilder.Authentication` property and relying on its behavior of automatically registering the AuthN and AuthZ middleware in the app, some puzzling behavior was observed.

In the following application, requests to `/swagger` are allowed, despite a fallback policy being set that requires authenticated users. Requests to the defined endpoints also show that the inline middleware is actually running before a 401 response is returned, which is unexpected as the implicitly added AuthZ middleware should have run before any middleware/endpoints defined in the app. The inline middleware does not run for requests to "/swagger" interestingly, suggesting that Swagger is possibly adding itself to the pipeline in a way that is exacerbating the behavior here.

Uncommenting the explicit `app.UseAuthorization()` line makes things run more like would be expected, with no requests getting through that stage to Swagger, the inline middleware, etc.

```c#
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.OpenApi;

var builder = WebApplication.CreateBuilder(args);

builder.Authentication.AddJwtBearer(); // <-- this will implicitly add AuthN/AuthZ middleware

builder.Services.AddAuthorization(authzOptions =>
    {
        var policy = new AuthorizationPolicyBuilder()
            .RequireAuthenticatedUser()
            .Build();
        authzOptions.DefaultPolicy = policy;
        authzOptions.FallbackPolicy = policy;
    });

builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

var app = builder.Build();

//app.UseAuthorization();

if (app.Environment.IsDevelopment())
{
    app.UseSwagger();
    app.UseSwaggerUI();
}

app.UseHttpsRedirection();

app.Use((ctx, next) =>
{
    Console.WriteLine($"Inline middleware ran. Current endpoint is '{ctx.GetEndpoint()?.DisplayName}'");
    return next();
});

app.MapGet("/", (HttpContext ctx) =>
{
    Console.WriteLine($"Route handler ran. Current endpoint is '{ctx.GetEndpoint()?.DisplayName}'");
    return "Site root";
});
app.MapGet("/hello", (HttpContext ctx) =>
{
    Console.WriteLine($"Route handler ran. Current endpoint is '{ctx.GetEndpoint()?.DisplayName}'");
    return "Hello, World!";
});

app.Run();
```

### PR DESCRIPTION
**Pull request by captainsafia**
Closes https://github.com/dotnet/aspnetcore/issues/42240.

### CODE REVIEW

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs [hunk 1]
```diff
@@ -1976,6 +1981,43 @@ public void CanObserveDefaultServicesInServiceCollection()
...
+        await using var app = builder.Build();
+
+        app.Use(next =>
+        {
+            return async context =>
+            {
+                // IAuthenticationFeature is added by the authentication middleware
+                // during invocation. This middleware should run after authentication
+                // and be able to access the feature.
+                var authFeature = context.Features.Get<IAuthenticationFeature>();
```
Comment (davidfowl) [inline src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:2002]: - Add a bool, to verify that this code path ran and check if the user is set? 
- I would rewrite this test to use a custom auth handler that makes the user authenticated when a header is set or something that's easy to set with the test server.
- Verify the user is set and the identity's name matches whatever you returned from the auth handler.

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs [hunk 2]
```diff
@@ -2001,21 +2007,46 @@ public async Task RegisterAuthMiddlewaresCorrectly()
...
+        {
+            var username = _context.Request.Query["username"];
+            var principal = new ClaimsPrincipal();
+            var id = new ClaimsIdentity();
+            id.AddClaim(new Claim(ClaimsIdentity.DefaultNameClaimType, username));
+            principal.AddIdentity(id);
+            return Task.FromResult(AuthenticateResult.Success(
+                new AuthenticationTicket(principal, "custom")));
+        }
+        public Task ChallengeAsync(AuthenticationProperties properties) => Task.FromResult(0);
```
Comment (davidfowl) [inline src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:2040]: nits:

```suggestion
        public Task ChallengeAsync(AuthenticationProperties properties) => Task.CompletedTask;
```

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs [hunk 3]
```diff
@@ -2001,21 +2007,46 @@ public async Task RegisterAuthMiddlewaresCorrectly()
...
+
+        public Task ForbidAsync(AuthenticationProperties properties) => Task.FromResult(0);
```
Comment (davidfowl) [inline src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:2042]:
```suggestion
        public Task ForbidAsync(AuthenticationProperties properties) => Task.CompletedTask;
```

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs [hunk 4]
```diff
@@ -1976,6 +1981,74 @@ public void CanObserveDefaultServicesInServiceCollection()
...
+        {
+            var username = _context.Request.Query["username"];
+            var principal = new ClaimsPrincipal();
+            var id = new ClaimsIdentity();
+            id.AddClaim(new Claim(ClaimsIdentity.DefaultNameClaimType, username));
+            principal.AddIdentity(id);
+            return Task.FromResult(AuthenticateResult.Success(
+                new AuthenticationTicket(principal, "custom")));
+        }
+        public Task ChallengeAsync(AuthenticationProperties properties) => Task.FromResult(0);
```
Comment (captainsafia) [inline src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:2040]:
```suggestion
        public Task ChallengeAsync(AuthenticationProperties properties) => Task.CompletedTask;
```

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs [hunk 5]
```diff
@@ -1976,6 +1981,74 @@ public void CanObserveDefaultServicesInServiceCollection()
...
+
+        public Task ForbidAsync(AuthenticationProperties properties) => Task.FromResult(0);
+
+        public Task<bool> HandleRequestAsync() => Task.FromResult(false);
```
Comment (captainsafia) [inline src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:2044]:
```suggestion
        public Task<bool> HandleRequestAsync() => Task.FromResult(false);

```

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs [hunk 6]
```diff
@@ -1976,6 +1981,74 @@ public void CanObserveDefaultServicesInServiceCollection()
...
+        public Task InitializeAsync(AuthenticationScheme scheme, HttpContext context)
+        {
+            _context = context;
+            return Task.FromResult(0);
```
Comment (captainsafia) [inline src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:2048]:
```suggestion
            return Task.CompletedTask;
```

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/Microsoft.AspNetCore.Tests.csproj [hunk 7]
```diff
@@ -7,6 +7,8 @@
   <ItemGroup>
     <Reference Include="Microsoft.AspNetCore" />
     <Reference Include="Microsoft.AspNetCore.TestHost" />
+    <Reference Include="Microsoft.AspNetCore.Authentication.JwtBearer" />
+    <Reference Include="Microsoft.AspNetCore.Authorization.Policy" />
```
Comment (davidfowl) [inline src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/Microsoft.AspNetCore.Tests.csproj:11]: REmove?

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs [hunk 8]
```diff
@@ -1976,6 +1981,75 @@ public void CanObserveDefaultServicesInServiceCollection()
...
+                // and be able to access the feature.
+                var authFeature = context.Features.Get<IAuthenticationFeature>();
+                Assert.NotNull(authFeature);
+                customMiddlewareExecuted = true;
+                Assert.Equal("foobar", context.User.Identity.Name);
+                await next(context);
+            };
+        });
+
+        app.MapGet("/hello", () => helloEndpointCalled = true).AllowAnonymous();
```
Comment (davidfowl) [inline src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs:2016]: Can you also verify the user name is set here too?

**Review by davidfowl — APPROVED** [review]

File: src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs (hunk 8 above)
Comment (captainsafia) [reply src/DefaultBuilder/test/Microsoft.AspNetCore.Tests/WebApplicationTests.cs]: Done.

**Review by Tratcher — APPROVED** [review]

**Pull request merged**